### PR TITLE
L1 regularization function

### PR DESCRIPTION
**Pull request by karangrewal**
The new file `nupic/research/frameworks/pytorch/regularization.py` contains `l1_regularization_step` which performs in place L1 penalty updates to specified parameters during training. I have also included a set of unit tests for this function.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/regularization.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        loss.backward()
+        optimizer.step()
+        l1_regularization_step(params=model.parameters(), lr=0.1, weight_decay=1e-3)
+
+
+    :param params: a list of parameters on which the L1 regularization update will be
+                   performed, conditioned on whether attribute `requires_grad` is True
+    :param lr: the learning rate used during optimization, analogous to the `lr`
+               parameter in `torch.optim.SGD`
+    :param weight_decay: the L1 penalty coefficient, analogous to the `weight_decay`
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/regularization.py:51]: It assumes the optimizer does not use `weight_decay`. Correct?

**Review by lscheinkman — APPROVED** [review]

File: nupic/research/frameworks/pytorch/regularization.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/regularization.py:51]: The optimizer's `weight_decay` only applies to the L2 penalty. It doesn't matter if the optimizer uses weight decay or not, since that doesn't affect the L1 penalty (and vice-versa)

File: nupic/research/frameworks/pytorch/regularization.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import torch
+
+
+@torch.no_grad()
+def l1_regularization_step(params, lr, weight_decay=1e-3):
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/regularization.py:26]: Is it based on any existing implementation? If yes, it is often useful to add the reference.

**Review by lucasosouza — APPROVED** [review]
Tests looks great, awesome. Left a minor comment.

File: nupic/research/frameworks/pytorch/regularization.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/regularization.py:26]: Thanks, I added the reference in the docstring.

**Comment by karangrewal** [comment]
Note: something seems wrong with CircleCI. It's giving flake8 errors on files that were not part of my commit, but those files never gave an issue before. See [here](https://app.circleci.com/pipelines/github/numenta/nupic.research/1495/workflows/e87df36a-1abb-40b1-b7aa-63aa496e323e/jobs/7447) for more info.

**Comment by lucasosouza** [comment]
> Note: something seems wrong with CircleCI. It's giving flake8 errors on files that were not part of my commit, but those files never gave an issue before. See [here](https://app.circleci.com/pipelines/github/numenta/nupic.research/1495/workflows/e87df36a-1abb-40b1-b7aa-63aa496e323e/jobs/7447) for more info.

flake8 versions are all fixed in requirements-dev.txt, so unless we updated that file recently it is strange this is happening. @lscheinkman do you have any ideas what may be causing this?

**Comment by lscheinkman** [comment]
`E741` is a real error. See https://www.python.org/dev/peps/pep-0008/#names-to-avoid
Somehow flake8 did not run during https://github.com/numenta/nupic.research/pull/356 and other PRs
We should definitely fix these issues before merging

**Comment by lucasosouza** [comment]
> `E741` is a real error. See https://www.python.org/dev/peps/pep-0008/#names-to-avoid
> Somehow flake8 did not run during #356 and other PRs
> We should definitely fix these issues before merging or create a separate PR fixing `E741`

But it is not related to this PR, we should create a separate JIRA and PR for this. It is strange that this was not caught before, or that is not showing up in other PRs that were merged yesterday and today. 

@mvacaporale you might want to see this, it is mainly related to naming variables in some code you wrote in the past.

**Comment by mvacaporale** [comment]
It's odd that so many errors came up all at once. It's not just the recent PR that was called out, but even so older code as well.

File: requirements.txt [hunk 3]
```diff
@@ -26,6 +26,7 @@ nupic.tensorflow @ git+https://github.com/numenta/nupic.tensorflow.git
 awscli
 ax-platform
 botorch
+colorama=0.4.3
```
Comment (lscheinkman) [inline requirements.txt:29]: It should be `colorama==0.4.3`
